File: actor.ts
import { Actor } from "./bsky.d.ts";
import { get_handle, get_did, sanitize } from "./utils.ts";

export async function get_actor (handle: string): Actor {
  let did;
  if (handle.includes("did:")) {
    did = handle;
    handle = await get_handle(handle);
  } else {
    did = await get_did(handle);
  }

  const res = await fetch(`https://api.bsky.app/xrpc/app.bsky.actor.getProfile?actor=${did}`)
    .then(res => res.json());

  if (res.error) return res;

  const record = res;

  let avatar, banner;

  if (record.avatar) {
    avatar = record.avatar ?
    record.avatar :
    "https://upload.wikimedia.org/wikipedia/commons/thumb/5/59/Minecraft_missing_texture_block.svg/2048px-Minecraft_missing_texture_block.svg.png";
  } else {
    avatar = "/resources/avatar.png";
  }

  if (record.banner) {
    banner = record.banner ?
    record.banner :
    "https://upload.wikimedia.org/wikipedia/commons/thumb/5/59/Minecraft_missing_texture_block.svg/2048px-Minecraft_missing_texture_block.svg.png";
  } else {
    banner = "/resources/banner.png";
  }

  const name = record.displayName ? sanitize(record.displayName) : handle;

  let description;
  if (record.description !== "") {
    description = sanitize(record.description);
  } else {
    description = "";
  }


  const actor: Actor = {
    avatar: avatar,
    banner: banner,
    name: name,
    description: description,
    handle: handle,
    follows: record.followsCount,
    followers: record.followersCount,
    posts: record.postsCount,
  };

  return actor;
}

File: actor.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"html/template"
	"io"
	"net/http"
)

type Actor struct {
	DID            string `json:"did"`
	Handle         string `json:"handle"`
	DisplayName    string `json:"displayName"`
	Description    string `json:"description"`
	Avatar         string `json:"avatar"`
	Banner         string `json:"banner"`
	FollowersCount int    `json:"followersCount"`
	FollowsCount   int    `json:"followsCount"`
	PostsCount     int    `json:"postsCount"`
	Feed           []FeedItem
}

func getActorProfile(did string) Actor {
	res, err := http.Get("https://api.bsky.app/xrpc/app.bsky.actor.getProfile?actor=" + did)
	if err != nil {
		fmt.Println(err)
	}

	var actor Actor
	b, _ := io.ReadAll(res.Body)
	json.Unmarshal(b, &actor)

	return actor
}

func getActorPage(actor Actor, feed Feed) string {
	t := template.Must(template.ParseFS(publicFiles, "public/*"))
	actor.Feed = feed
	var actor_page bytes.Buffer
	t.ExecuteTemplate(&actor_page, "actor.page.html", actor)
	return actor_page.String()
}
